Frameworks/QuartzCore/CALayer.mm

```diff
@@ -518,69 +521,73 @@ - (void)display {
+            // UIKit and CALayer consumers expect the origin to be in the top left.
+            // CoreGraphics defaults to the bottom left, so we must flip and translate the canvas.
+            CGContextTranslateCTM(drawContext, 0, heightInPoints);
+            CGContextScaleCTM(drawContext, 1.0f, -1.0f);
+            CGContextTranslateCTM(drawContext, -priv->bounds.origin.x, -priv->bounds.origin.y);
 
-            CGRect wholeRect = CGRectMake(0, 0, width, height);
-            CGContextFillRect(drawContext, wholeRect);
-            CGContextRestoreGState(drawContext);
-        }
+            _CGContextSetShadowProjectionTransform(drawContext, CGAffineTransformMakeScale(1.0, -1.0));
 
-        // UIKit and CALayer consumers expect the origin to be in the top left.
-        // CoreGraphics defaults to the bottom left, so we must flip and translate the canvas.
-        CGContextTranslateCTM(drawContext, 0, heightInPoints);
-        CGContextScaleCTM(drawContext, 1.0f, -1.0f);
-        CGContextTranslateCTM(drawContext, -priv->bounds.origin.x, -priv->bounds.origin.y);
+            [self drawInContext:drawContext];
 
-        _CGContextSetShadowProjectionTransform(drawContext, CGAffineTransformMakeScale(1.0, -1.0));
+            if (priv->delegate != 0) {
+                if ([priv->delegate respondsToSelector:@selector(displayLayer:)]) {
+                    [priv->delegate displayLayer:self];
+                } else {
+                    [priv->delegate drawLayer:self inContext:drawContext];
+                }
+            }
 
-        CGContextSetDirty(drawContext, false);
-        [self drawInContext:drawContext];
+            _CGContextPopEndDraw(drawContext);
 
-        if (priv->delegate != 0) {
-            if ([priv->delegate respondsToSelector:@selector(displayLayer:)]) {
-                [priv->delegate displayLayer:self];
-            } else {
-                [priv->delegate drawLayer:self inContext:drawContext];
+            woc::StrongCF<CFErrorRef> renderError;
+            if (CGContextIwGetError(drawContext, &renderError)) {
+                switch (CFErrorGetCode(renderError)) {
+                    case kCGContextErrorDeviceReset:
+                        NSTraceInfo(TAG, @"Hardware device disappeared when rendering %@; retrying.", self);
+                        ++tries;
+                        continue;
+                    default: {
+                        NSTraceError(TAG, @"Error %@ rendering %@.", renderError.get(), self);
+                        break;
+                    }
+                }
             }
-        }
 
-        CGContextReleaseLock(drawContext);
+            CGImageRef target = CGBitmapContextGetImage(drawContext);
+            priv->ownsContents = TRUE;
+            priv->savedContext = CGContextRetain(drawContext);
+            priv->contents = CGImageRetain(target);
+            break;
+        } while (tries < 3);
```

Comment:
I don't know about exposing it, but I think giving it a name/placing it next to the declaration of tries might be slightly more readable.